proof_ipa.go

```diff
@@ -102,18 +117,28 @@ func (sd StateDiff) Copy() StateDiff {
 	ret := make(StateDiff, len(sd))
 	for i := range sd {
 		copy(ret[i].Stem[:], sd[i].Stem[:])
-		ret[i].SuffixDiffs = make([]SuffixStateDiff, len(sd[i].SuffixDiffs))
-		for j := range sd[i].SuffixDiffs {
-			ret[i].SuffixDiffs[j].Suffix = sd[i].SuffixDiffs[j].Suffix
-			if sd[i].SuffixDiffs[j].CurrentValue != nil {
-				ret[i].SuffixDiffs[j].CurrentValue = &[32]byte{}
-				copy((*ret[i].SuffixDiffs[j].CurrentValue)[:], (*sd[i].SuffixDiffs[j].CurrentValue)[:])
-			}
-			if sd[i].SuffixDiffs[j].NewValue != nil {
-				ret[i].SuffixDiffs[j].NewValue = &[32]byte{}
-				copy((*ret[i].SuffixDiffs[j].NewValue)[:], (*sd[i].SuffixDiffs[j].NewValue)[:])
-			}
+
+		ret[i].Updates = make([]UpdateDiff, len(sd[i].Updates))
+		for j := range sd[i].Updates {
+			ret[i].Updates[j].Suffix = sd[i].Updates[j].Suffix
+			ret[i].Updates[j].Current = sd[i].Updates[j].Current
```

Comment:
Since we stop having `[]byte`, many of `if len(xxx) == 0` cases and similar are gone. We can always copy-by-value.